### PR TITLE
TLS support, connection options & some refactoring

### LINKED ISSUE #18: Add TLS database connection support

### PR DESCRIPTION
**Pull request by EinKrebs**
- Added database name, username and password options for connection to database;
- abled connection using TLS;
- did some code refactoring;
- changed README.

Can you please give me code review and some ideas about README content?

Resolves #18

### CODE REVIEW

**Comment by EinKrebs** [comment]
Also:
- Fixed span finding issue(check commit message for details);
- Changed run options to make it more universal.

File: storage/config.go [hunk 1]
```diff
@@ -20,4 +20,32 @@ type Configuration struct {
...
+		cfg.BatchFlushInterval = defaultBatchDelay
+	}
+	if cfg.Encoding == "" {
+		cfg.Encoding = JSONEncoding
+	}
+	if cfg.Username == "" {
+		cfg.Username = "default"
+	}
+	if cfg.Database == "" {
+		cfg.Database = "default"
```
Comment (pavolloffay) [inline storage/config.go:49]: does this "default" database has to be explicitly created (e.g. when running from docker)?

File: storage/store.go [hunk 2]
```diff
@@ -130,8 +135,38 @@ func (s *Store) Close() error {
 	return s.db.Close()
 }
 
-func defaultConnector(datasource string) (*sql.DB, error) {
-	db, err := sql.Open("clickhouse", datasource)
+func defaultConnector(cfg Configuration) (*sql.DB, error) {
```
Comment (pavolloffay) [inline storage/store.go:138]: I don't like having `defaultConnector` and `tlsConnector` especially when both accept full Configuration object. The caller has to know which one to call based on the information in the Configuration. 

```
func createDB(cfg Configuration) (*sql.DB, error) {
  connectionParams := fmt.Sprintf("%s?database=%s&username=%s&password=%s", cfg.Address, cfg.Database, cfg.Username, cfg.Password	)

  if cf.TLS {
    // add secure to connectionParams and load the certificate.
  }
```

File: README.md [hunk 3]
```diff
@@ -5,12 +5,34 @@ Jaeger ClickHouse gRPC [storage plugin](https://github.com/jaegertracing/jaeger/
 This is WIP and it is based on https://github.com/bobrik/jaeger/tree/ivan/clickhouse/plugin/storage/clickhouse. 
 See as well [jaegertracing/jaeger/issues/1438](https://github.com/jaegertracing/jaeger/issues/1438) for ClickHouse plugin.
 
+## Connection to remote database
+
+You can specify connection settings in ``config.yaml`` file
+
+### Using TLS connection
+
+For TLS connection, you need to put CA-certificate to project directory and specify path to it in ``config.yaml``
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply README.md:14]: Alternatively can someone just specify user and pass without cacert?

**Comment by pavolloffay** [comment]
@EinKrebs nit: generally try to keep changes in the PR minimal and rather split the PR into smaller PRs when doing multiple changes that do not belong together.

File: Makefile [hunk 4]
```diff
@@ -11,11 +11,11 @@ build:
 
 .PHONY: run
 run:
-	docker run --rm --name jaeger -e JAEGER_DISABLED=true --link some-clickhouse-server -it -u ${shell id -u} -p16686:16686 -p14250:14250 -p14268:14268 -p6831:6831/udp -v "${PWD}:/data" -e SPAN_STORAGE_TYPE=grpc-plugin jaegertracing/all-in-one:${JAEGER_VERSION} --query.ui-config=/data/jaeger-ui.json --grpc-storage-plugin.binary=/data/jaeger-clickhouse-$(GOOS)-$(GOARCH) --grpc-storage-plugin.configuration-file=/data/config.yaml --grpc-storage-plugin.log-level=debug
+	docker run --rm --name jaeger -e JAEGER_DISABLED=true -it -u ${shell id -u} --network="host" -p16686:16686 -p14250:14250 -p14268:14268 -p6831:6831/udp -v "/usr/share/zoneinfo/:/usr/share/zoneinfo" -v "${PWD}:/data" -e SPAN_STORAGE_TYPE=grpc-plugin jaegertracing/all-in-one:${JAEGER_VERSION} --query.ui-config=/data/jaeger-ui.json --grpc-storage-plugin.binary=/data/jaeger-clickhouse-$(GOOS)-$(GOARCH) --grpc-storage-plugin.configuration-file=/data/config.yaml --grpc-storage-plugin.log-level=debug
```
[the remark this answers is not in the document]
Comment (EinKrebs) [reply Makefile:14]: Troubles with connection to ClickHouse DB with non-UTC timezone. Storage couldn't find timezone info, so it crashed.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (EinKrebs) [reply README.md:14]: > is username and password required as well for TLS?

No

> Alternatively can someone just specify user and pass without cacert?

Yes, and also database name

File: storage/config.go (hunk 1 above)
Comment (EinKrebs) [reply storage/config.go:49]: No

**Comment by pavolloffay** [comment]
@EinKrebs PR needs to be rebased

**Comment by EinKrebs** [comment]
@pavolloffay done.

**Comment by pavolloffay** [comment]
It seems that there are some issue 

```
This branch cannot be rebased due to conflicts
Rebasing the commits of this branch on top of the base branch cannot be performed automatically due to conflicts encountered while reapplying the individual commits from the head branch.
```

**Comment by pavolloffay** [comment]
I am not sure what is the issue but I have also noticed that you are submitting the PR from your main branch. I would suggest using feature branches and keep main in sync with the upstream main branch.

**Comment by EinKrebs** [comment]
Thank you for advice, it's really much easier to work like this.

File: storage/store.go [hunk 5]
```diff
@@ -39,24 +39,19 @@ var _ shared.ArchiveStoragePlugin = (*Store)(nil)
 var _ io.Closer = (*Store)(nil)
 
 func NewStore(logger hclog.Logger, cfg Configuration, embeddedSQLScripts embed.FS) (*Store, error) {
-	db, err := defaultConnector(cfg.Address)
+	cfg.setDefaults()
+	var db *sql.DB
+	var err error
+	db, err = connector(cfg)
```
Comment (pavolloffay) [inline storage/store.go:45]: why is `db` and `err` defined above? They can be defined directly here as `db, err := connector(cfg)`

**Review by pavolloffay — COMMENTED** [review]
Please rebase the PR It cannot be merged as it is.

**Comment by EinKrebs** [comment]
I don't know why branch can't be rebased, when I try to rebase it my git does nothing

**Comment by EinKrebs** [comment]
I made better version of this PR.

**Pull request closed without merging**